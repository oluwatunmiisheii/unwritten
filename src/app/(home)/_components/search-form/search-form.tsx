'use client'
import { useRouter } from 'next/navigation'

export const SearchForm = () => {
  const router = useRouter()

  async function search(formData: FormData) {
    const searchQuery = formData.get('search') as string
    router.push(`/?search=${encodeURIComponent(searchQuery)}`, { scroll: false })
  }

  return (
    <form className="flex gap-x-4" action={search}>
      <input
        id="search"
        name="search"
        type="text"
        required
        aria-label="Search"
        className="block w-full rounded-md bg-white px-3 py-1.5 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 placeholder:text-gray-400 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-blue-500 sm:text-sm/6"
      />
      <button
        type="submit"
        className="rounded-md bg-blue-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600 flex-shrink-0"
      >
        Search
      </button>
    </form>
  )
}
