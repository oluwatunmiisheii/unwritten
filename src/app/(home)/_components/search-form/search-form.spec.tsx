import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import mockRouter from 'next-router-mock'
import { expect, test, describe, vi } from 'vitest'
import { SearchForm } from './search-form'
import userEvent from '@testing-library/user-event'

vi.mock('next/navigation', () => vi.importActual('next-router-mock'))

describe('SearchForm', () => {
  test('renders as expected', async () => {
    const { container } = render(<SearchForm />)

    expect(screen.getByLabelText(/search/i)).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /search/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  test('calls router.push with the correct URL when form is submitted', async () => {
    const user = userEvent.setup()
    render(<SearchForm />)

    const searchInput = screen.getByLabelText(/search/i)
    const searchButton = screen.getByRole('button', { name: /search/i })

    await user.type(searchInput, 'test query')
    await user.click(searchButton)

    expect(mockRouter).toMatchObject({
      asPath: '/',
      pathname: '/',
      query: { search: 'test query' },
    })
  })

  test('prevents form submission when input is empty', async () => {
    const user = userEvent.setup()
    render(<SearchForm />)

    const searchButton = screen.getByRole('button', { name: /search/i })

    await user.click(searchButton)

    expect(mockRouter).toMatchObject({
      asPath: '/',
      pathname: '/',
      query: {},
    })
  })
})
