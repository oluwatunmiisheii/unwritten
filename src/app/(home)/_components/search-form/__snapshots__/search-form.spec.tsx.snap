// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders as expected 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="flex gap-x-4"
  >
    <input
      aria-label="Search"
      class="block w-full rounded-md bg-white px-3 py-1.5 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 placeholder:text-gray-400 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-blue-500 sm:text-sm/6"
      id="search"
      name="search"
      required=""
      type="text"
    />
    <button
      class="rounded-md bg-blue-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600 flex-shrink-0"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
